### ISSUE TITLE
feature request: a way to globally override the git protocol

### ISSUE THREAD

**Issue by kernelsmith** [berkshelf/berkshelf#1503]
I find in most environments that the clear-text git protocol is blocked (port 9418), so if a Berksfile has indicated github: 'some/repo' at any point, I'm going to have to fork and maintain my own so I can switch to the long form git: option to utilize ssh or https.  The github: option seems to have the url handler hard coded at https://github.com/berkshelf/berkshelf/blob/master/lib/berkshelf/locations/github.rb#L4  If we made the handler dynamic we could let the user specify it and of course default it to the git protocol.

I'm perfectly willing to submit a pull request for this functionality, but I'd like to get some comments from committers on both the "if we should do this" and "how should we do this" fronts.  I'm perfectly willing to come up w/my own "how" but I've been burned on this before when someone's not on board w/the "if" and/or the "how".  At a cursory glance, it would seem ideal to have something in .berkshelf/config.json, maybe git.handler or git.url or git.protocol.

Again, happy to submit a PR, but I've never submitted to berkshelf, however I have submitted to other ruby projects and was a developer (commit rights) on Metasploit for a while

**Comment by reset** [comment]
@kernelsmith it sounds perfectly reasonable that you could add a configuration option to berkshelf config file to choose between using the git, https or ssh protocol. If you want to tackle that and provide a PR that exposes this option while keeping the default behaviour the same as we already have I'll review it and merge for you

**Comment by kernelsmith** [comment]
awesome @reset thanks much for your time

**Comment by kernelsmith** [comment]
cool, it's working (https://github.com/kernelsmith/berkshelf/tree/feature/1503-global-git-handler).  I'll check what, if anything, should change in the specs.  I didn't see anywhere in buff-config or varia_model where I could validate/restrict the new config option (github.transport) to values such as "git", "ssh", "https" specifically, just to the type String.  I also didn't see a great place to do it in berkshelf itself, so instead I just treat "ssh" and "https" especially and anything and everything else is treated as "git"

**Comment by kernelsmith** [comment]
need to change the config option tho, github.transport will conflict w/the existing github option which is an array.  I'm thinking github_transport?

**Comment by reset** [comment]
@kernelsmith let's go with `github.protocol`

**Comment by reset** [comment]
Additionally, the three options should be `:ssh`, `:https`, and `:git`. The `:git` option should be the default.

**Comment by kernelsmith** [comment]
no problem, but won't we have a problem w/the json in the config file because of the existing github option?  Currently this is the case right:

```json
{
  "github": [
    {"someuser":"somepass"},
    {"otheruser":"otherpass"}
  ]
}
```
And we want to use that "namespace" like this:

```json
{
  "github": {
    "protocol": "ssh"
  }
}
```
We could alter the github item to accommodate both, but that would require some additional changes, but could have something like:

```json
{
  "github": {
    "creds": [
      {"someuser":"somepass"},
      {"otheruser":"otherpass"}
    ],
    "protocol":"ssh"
  }
}
```
I admit I'm not great w/json + buff::config + varia_model

**Comment by kernelsmith** [comment]
@reset could you advise?

**Comment by reset** [comment]
@kernelsmith I'm pretty sure we whitelist the options passed around so you shouldn't need to worry about using this form:

```json
{
  "github": {
    "username": "reset",
    "password": "mypass",
    "protocol": "ssh",
  }
}
```

Of course, if the tests don't pass, then we need to change the name of the config option ;)

**Comment by kernelsmith** [comment]
@reset, ok.  Usage says github should be an array of credentials to try, tho what you have shown is a hash.  If the documentation is indeed right, I would think the tests would fail, tho I don't know exactly what the whitelisting processing is doing, so I'll just give that a shot and run the tests and see what, if anything, explodes

**Comment by kernelsmith** [comment]
fighting w/the old 
`ERROR - Invalid Guardfile, original error is:
> [#] wrong number of arguments (2 for 0..1)` guard error atm.

**Comment by kernelsmith** [comment]
I haven't lost this, gonna create a fresh dev env to see if I can get guard to do the right thing

**Comment by kernelsmith** [comment]
sorry, had too many other things going on, I'm back to setting up a new dedicated dev env

**Comment by kernelsmith** [comment]
any ideas here?

```
$ cd gits/berkshelf
~/gits/berkshelf$ bundle install
...snip... (all good)
~/gits/berkshelf$ BERKSHELF_DEBUG=1 bundle exec guard start
22:35:43 - ERROR - Invalid Guardfile, original error is:
> [#] wrong number of arguments (2 for 0..1)
22:35:43 - INFO - Guard is using Libnotify to send notifications.
22:35:43 - INFO - Guard is using TerminalTitle to send notifications.
22:35:43 - INFO - Starting Spork for RSpec, Cucumber
Using Cucumber
Using RSpec
Loading Spork.prefork block...
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:35: warning: toplevel constant Cucumber referenced by Spork::Server::Cucumber
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:35:in `rescue in preload_legacy_cucumbers'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:31:in `preload_legacy_cucumbers'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:15:in `preload'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/run_strategy/forking.rb:25:in `preload'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/runner.rb:74:in `run'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/lib/spork/runner.rb:10:in `run'
/home/user/.rvm/gems/ruby-2.1.8/gems/spork-0.9.2/bin/spork:10:in `<top (required)>'
/home/user/.rvm/gems/ruby-2.1.8/bin/spork:23:in `load'
/home/user/.rvm/gems/ruby-2.1.8/bin/spork:23:in `<main>'
/home/user/.rvm/gems/ruby-2.1.8/bin/ruby_executable_hooks:15:in `eval'
/home/user/.rvm/gems/ruby-2.1.8/bin/ruby_executable_hooks:15:in `<main>'
Spork is ready and listening on 8989!
22:45:58 - ERROR - Could not start Spork server for RSpec, Cucumber after 30 seconds. I will continue waiting for a further 60 seconds.
```
I'm running rvm 1.27.0, and the following gem versions (picked a subset of what I thought might be pertinent):
cucumber (2.3.3, 2.3.2)
cucumber-core (1.4.0)
cucumber-wire (0.0.1)
guard-cucumber (2.0.0)
berkshelf-api-client (2.0.2, 2.0.0)
guard (1.8.3)
guard-compat (1.2.1)
guard-rspec (1.2.2)
guard-spork (1.5.2)

**Comment by kernelsmith** [comment]
ok, I give up on getting the tests to run, blown about 5x more time trying to get this to work than I did writing the code.  I'm gonna submit the PR as is.

```
zdi@ubuntu:~/gits/berkshelf$ git branch
* feature/1503-global-git-handler
master
zdi@ubuntu:~/gits/berkshelf$ BERKSHELF_DEBUG=1 bundle exec guard start
17:02:17 - ERROR - Invalid Guardfile, original error is:
> [#] wrong number of arguments (given 2, expected 0..1)
17:02:17 - INFO - Guard is using Libnotify to send notifications.
17:02:17 - INFO - Guard is using TerminalTitle to send notifications.
17:02:17 - INFO - Starting Spork for RSpec, Cucumber
Using Cucumber
Using RSpec
Loading Spork.prefork block...
/home/zdi/.rvm/gems/ruby-2.3.0/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:35: warning: toplevel constant Cucumber referenced by Spork::Server::Cucumber
undefined method `mother_object' for Cucumber:Module (NoMethodError)
/home/zdi/.rvm/gems/ruby-2.3.0/gems/spork-0.9.2/lib/spork/test_framework/cucumber.rb:35:in `rescue in preload_legacy_cucumbers'
...
```

**Comment by reset** [comment]
@kernelsmith you don't need to use Guard to run the tests, you can just run them with `bundle exec thor spec:ci`

**Comment by kernelsmith** [comment]
Ah, contributing.md makes no mention of that, only guard

**Comment by kernelsmith** [comment]
Fixed in #1550
